update anova type II and III

clarkliming opened the pull request:
close #392

danielinteractive commented:
Thanks @clarkliming , can you add a new issue for this production work? (ideally also with checkboxes for the objectives of that issue) The other issue was only for the design I think

danielinteractive reviewed (commented):
Thanks @clarkliming , looks great!
Few suggestions:
- Could we use a similar mechanism as for `emmeans` so that we don't need to import `car` but just suggest it?
- could we add a `NEWS` item
- For the vignette I remember we somewhere have something on hypothesis testing already, would it make sense to integrate those together somehow?
- please add an example in introduction vignette

clarkliming commented:
> Thanks @clarkliming , looks great! Few suggestions:
> 
> * Could we use a similar mechanism as for `emmeans` so that we don't need to import `car` but just suggest it?
Updated. However added a new function (copy from emmeans) but it is difficult to do any testing on it

> * could we add a `NEWS` item
Already had an entry under "new features". do we need more?

> * For the vignette I remember we somewhere have something on hypothesis testing already, would it make sense to integrate those together somehow?

Is it the section in introduction? Maybe it is clearer to be separated here because it involves a lot of details?

> * please add an example in introduction vignette

Added now

clarkliming commented on R/mmrm-methods.R:
```diff
@@ -20,6 +20,10 @@
 #'    2. If the covariance structure is spatial, the covariance matrix of two time points with unit distance
 #'       will be displayed.
 #'
+#' `Anova` will return `anova` object with one row per variable and columns
+#' `Num Df`(numerator degrees of freedom), `Denom Df`(denominator degrees of freedom),
+#' `F Statistic` and `Pr(>=F)`.
+#'
```
I don't want to export `Anova.mmrm` so no place to properly document it. so still put it here.

clarkliming commented on R/zzz.R:
```diff
@@ -15,6 +15,11 @@
     message = emit_tidymodels_register_msg
   )
+  register_on_load(
+    "car", c("3.1.2", NA),
```
the version can be very flexible but let's use latest version

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+Before we discuss the testing, it is important that we understand the concept of "contained effect".
+The definition of "contained effect" is:
+
+For effect "E2", it is said to contain "E1" if
```
Can we rather use $E_2$ etc. that is more mathematical notation?

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+For effect "E2", it is said to contain "E1" if
+1. they both involve the same continuous variable (and no more other continuous variable is involved)
+1. "E2" have equal or more "categorical variable" than "E1"
```
```suggestion
1. "E2" have equal or more categorical covariates than "E1"
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+The definition of "contained effect" is:
+
+For effect "E2", it is said to contain "E1" if
+1. they both involve the same continuous variable (and no more other continuous variable is involved)
```
```suggestion
1. they both involve the same continuous covariates
```
I guess we don't need the parenthesis? Or we need to reword to make it clearer?

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+1. they both involve the same continuous variable (and no more other continuous variable is involved)
+1. "E2" have equal or more "categorical variable" than "E1"
+
+For example, for the following formula
```
```suggestion
For example, for the following model formula
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+we have
+
+1. If $A$ and $B$ are both categorical, then $F_{A*B}$ contains $F_{A}$ and $F_{B}$.
```
please first define what $F_{X}$ means, otherwise this is not clear

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+$X_1$ stands for columns of $X$ associated with $F_1$,
+$X_2$ stands for columns of $X$ whose associated effect contains $F_1$,
+$M = I - X_0(X_0^\top X_0)^{-}X_0^\top$,
+and $Z^{-}$ stands for the g2 inverse of $Z$.
```
```suggestion
and $Z^{-}$ stands for the g2 inverse of $Z$.
```
please include a link to a g2 inverse definition (I don't think this is commonly known)

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+$M = I - X_0(X_0^\top X_0)^{-}X_0^\top$,
+and $Z^{-}$ stands for the g2 inverse of $Z$.
+
+Note: In `mmrm` package, we do not allow singularity in general, so the g2 inverse is the inverse.
```
```suggestion
Note: Here we do not allow for singularity in general, so the g2 inverse is just the usual inverse.
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+and $Z^{-}$ stands for the g2 inverse of $Z$.
+
+Note: In `mmrm` package, we do not allow singularity in general, so the g2 inverse is the inverse.
+Thus we can use identity matrix in step 2.
```
```suggestion
Thus we can use an identity matrix as the submatrix in step 2.
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+For type III hypothesis testing, we assume that the hypothesis tested should be the same for all
+designs with the same general form of estimable functions.
+It can be defined with the following step:
```
```suggestion
It can be defined with the following steps:
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+Note: In `mmrm` package, we do not allow singularity in general, so the g2 inverse is the inverse.
+Thus we can use identity matrix in step 2.
+
```
can you add an example here? I think that would be helpful

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+Here equate the values means that, given that the coefficients in step 2 are all 1 (diagonal), in step 3
+this 1 is divided by the total number of new levels.
+
```
also here an example will be great

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+there can be issues with the comparison between SAS and R, as in R we still use the first level as our reference.
+
+When we conduct the hypothesis testing, we will also face this issue.
+The result will be slightly different because of different reference groups, however, if correct contrast is used
```
but is it possible always to match results by custom options for reference levels? or not?
would be good to make that clear

clarkliming replied:
I think it is already clear that `if correct contrast is used we will get closer result`. (I used closer because anyway we will not have identical results)

However, we do not include the algorithm to create the correct reference.

danielinteractive replied:
Hm it is not clear to me yet :-) Is the way we show the code currently getting us "closest" to SAS results? or is there another way to get even closer?
If there is an algorithm to create the correct reference/contrast then let's include it

clarkliming replied:
well this is not "correct reference" but user's choices. e.g. for RACE we really don't need to say "which reference is correct, Aisan or White"? they essentially try to answer the same question, however the result can be a little different. For type II it will be a little complicated to obtain the contrast for the with different reference level, for type III it is easier, e.g. if there are three levels, using lvl1 as reference, we are testing `lvl2 - lvl1` and `lvl3 - lvl1` are both zero, and the contrast should be
```
1 0
0 1
```

however if we are using `lvl3` as reference, then we should test `lvl1 - lvl3`, `lvl2 - lvl3`. However in R we have `lvl2-lvl1`, `lvl3 - lvl1`. With a linear combination of these we can use

```
0 1
-1 -1
```
So here I use "correct contrast" to say that, e.g. in SAS we are testing `lvl1 - lvl3`, `lvl2 - lvl3`, in R we shoud not use identity matrix if we only have `lvl2 - lvl1` and `lvl3 - lvl1` and identity matrix, but use the previous form of matrix to correctly test `lvl1 - lvl3`, `lvl2 - lvl3`. It is not related to "correct reference", but "correct contrast" to reflect the same testing.

However a better solution is to use a different model to create the same design matrix.

In summary, I think we should not provide such functionality to update the contrast afterwards

danielinteractive replied:
ok makes sense, can you add this example to the vignette? I think that helps the user

clarkliming replied:
updated

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+For type II/III testing, it is true that the order of covariate should not affect the result.
+However, if there is an interaction term, the reference level can be changed.
+With different reference level the result will be slightly different.
```
```suggestion
With different reference levels the result will be slightly different.
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
````diff
@@ -0,0 +1,129 @@
+Then it is equivalent to the following `mmrm` model
+
+```
+FEV1 ~ ARMCD * SEX + ARMCD * FEV1_BL - FEV1_BL
````
```suggestion
FEV1 ~ ARMCD * SEX + ARMCD:FEV1_BL
```
would be simpler

clarkliming replied:
this is on purpose to use `ARMCD*FEV1_BL` to better illustrate the difference: in SAS using `ARMCD*FEV1_BL` will only add the interaction term, but not the two main effect.

danielinteractive replied:
can we then add a comment that in R you can do the "ARMCD:FEV1_BL" instead?

clarkliming replied:
added

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -0,0 +1,129 @@
+
+Because SAS will not include the covariate `FEV1_BL` by default, unless manually added.
+However in R we will add `FEV1_BL` by default.
+Please note that in this example we excludes the covariance structure part.
```
```suggestion
Please note that in this example we exclude the covariance structure part.
```

danielinteractive reviewed (commented):
Thank you @clarkliming ! Great stuff. I gave it a thorough last round of review.

clarkliming commented on R/interop-car.R:
```diff
@@ -0,0 +1,109 @@
+  assert_double(tol, finite = TRUE, len = 1L)
+  type <- match.arg(type)
+  vars <- colnames(attr(terms(mod$formula_parts$model_formula), "factors"))
+  ret <- lapply(
```
slightly safer but causes more issues later. vapply always simplifies the result to an array/matrix however we return a list. The mode is a list and this is not we expected. We will need to further update the code to create a normal data frame

danielinteractive commented on _pkgdown.yml:
```diff
@@ -33,18 +33,18 @@ articles:
       - covariance
-  - title: Degrees of Freedom
-    navbar: Degrees of Freedom
+  - title: Degrees of Freedom and Inference
+    navbar: Degrees of Freedom and Inference
```
```suggestion
  - title: Degrees of Freedom and Testing
    navbar: Degrees of Freedom and Testing
```
(inference is more estimating parameters etc)

danielinteractive commented on inst/WORDLIST:
```diff
@@ -1,3 +1,4 @@
+ARMCD
```
maybe set in backticks to avoid the need for this?

clarkliming replied:
it is in the matrix part so adding backtick is ugly and really not necessary


```math
\begin{bmatrix}
\mu & ARMCD & RACE & RACE & ARMCD:RACE & ARMCD:RACE\\
0 & 0 & 1 & 0 & 0 & 0\\ 
0 & 0 & 0 & 1 & 0 & 0\\ 
\end{bmatrix}
```

```math
\begin{bmatrix}
\mu & `ARMCD` & `RACE` & `RACE` & `ARMCD`:`RACE` & `ARMCD`:`RACE`\\
0 & 0 & 1 & 0 & 0 & 0\\ 
0 & 0 & 0 & 1 & 0 & 0\\ 
\end{bmatrix}
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -57,58 +55,207 @@ For type II hypothesis testing, the contrasts can be obtained through the follow
+\end{matrix}
+$$
+
+Please note that this is $2\times 6$ matrix, the rows is the number of coefficients that is associated with `RACE`,
```
```suggestion
Please note that this is a $2\times 6$ matrix, the rows is the number of coefficients that is associated with `RACE`,
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -57,58 +55,207 @@ For type II hypothesis testing, the contrasts can be obtained through the follow
+$$
+
+Please note that this is $2\times 6$ matrix, the rows is the number of coefficients that is associated with `RACE`,
+and the columns is the total number of coefficients `r length(coef(fit))`
```
```suggestion
and the columns is the total number of coefficients `r length(coef(fit))`.
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -57,58 +55,207 @@ For type II hypothesis testing, the contrasts can be obtained through the follow
+Please note that this is $2\times 6$ matrix, the rows is the number of coefficients that is associated with `RACE`,
+and the columns is the total number of coefficients `r length(coef(fit))`
+
+Then following step 2, we filled in identity matrix in the corresponding submatrix
```
```suggestion
Then following step 2, we filled in the identity matrix in the corresponding submatrix
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -57,58 +55,207 @@ For type II hypothesis testing, the contrasts can be obtained through the follow
+\end{matrix}
+$$
+
+Then following step 2, we filled in identity matrix in the corresponding submatrix
```
```suggestion
Then following step 2, we filled in the identity matrix in the corresponding submatrix
```

danielinteractive commented on vignettes/hypothesis_testing.Rmd:
```diff
@@ -57,58 +55,207 @@ For type II hypothesis testing, the contrasts can be obtained through the follow
+while in R we are testing `VIS2 - VIS1`, `VIS3 - VIS1` and `VIS4 - VIS1`.
+
+To correctly test `VIS1 - VIS4`, `VIS2 - VIS4` and `VIS3 - VIS4` in R, we need to update the contrast
+to the following matrix(ignoring other part) to get closer result.
```
```suggestion
to the following matrix (ignoring the other part) to get a closer result.
```

danielinteractive reviewed (approved):
Congrats @clarkliming , good to merge after last few suggestions are in!
